Fix HF loading when there's no metadata file to work with fsspec

ankitageorge opened the pull request:
Summary: HF loading when there is no metadata is an edge case for some users. We were previously calling safe_open(filename) to get the keys in the safetensors file, but this doesn't work with fsspec, when models have a different backend than local fs (ie. hf, s3 etc). This diff updates to open the file with fsspec.open() and then safetensors.deserialize() to get the keys

Test Plan: unit test and e2e test reading from hf

Differential Revision: D74181513




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

joecummings commented on torch/distributed/checkpoint/_hf_storage.py:
```diff
@@ -225,22 +225,24 @@ def read_metadata(self) -> Metadata:
-                    state_dict_metadata[k] = BytesStorageMetadata()
-                    storage_data[k] = safetensors_files[0]
+            with self.fs.create_stream(safetensors_files[0], "rb") as f:
+                deserialized = deserialize(f.read())
```
This will load the whole file into memory right now, yes?

I think it's the quickest fix so that's fine, but I'd love to know if there's a way in which we can just not use a metadata file here since it's only a single file.

ankitageorge replied:
according to the documentation it's lazy loading https://github.com/huggingface/safetensors/blob/dfd1b9f5378e3ac9ff3e85f562d1a074bab2a81d/bindings/python/py_src/safetensors/__init__.pyi#L3

ankitageorge replied:
I just took the documentation's word for it, but I'll try to think of if there's a better way

joecummings replied:
FWIW, it looks like deserialize is loading the entire file into memory immediately: 

<details><summary>Repro script</summary>
<p>

```python
import os
import time

import psutil

from safetensors import deserialize
from safetensors.torch import load_file


def memory_usage():
    # Get the current memory usage
    process = psutil.Process(os.getpid())
    memory_usage = process.memory_info().rss / (1024 * 1024)
    return memory_usage


# Check time and memory usage of ``load_file``
before = memory_usage()
start = time.time()
data = load_file("Qwen2-0.5B-Instruct/model.safetensors")
print(f"Time to call 'load_file': {time.time() - start}")
after = memory_usage()
print(f"Memory usage after: {after} MB")
print(f"Memory usage difference: {after - before} MB")

del data

# Check time and memory usage of ``deserialize``
before = memory_usage()
start = time.time()
with open("Qwen2-0.5B-Instruct/model.safetensors", "rb") as f:
    data = deserialize(f.read())
print(f"Time to call 'deserialize': {time.time() - start}")
after = memory_usage()
print(f"Memory usage before: {before} MB")
print(f"Memory usage difference: {after - before} MB")

del data

# Check time and memory usage of ``load_file`` with reading keys of the safetensors file
before = memory_usage()
start = time.time()
data = load_file("Qwen2-0.5B-Instruct/model.safetensors")
all_keys = {}
for key in data.keys():
    all_keys[key] = data[key]
print(f"Time to call 'load_file w/ reading keys': {time.time() - start}")
after = memory_usage()
print(f"Memory usage after: {after} MB")
print(f"Memory usage difference: {after - before} MB")

del data

# Check time and memory usage of ``deserialize`` with reading keys of the safetensors file
before = memory_usage()
start = time.time()
with open("Qwen2-0.5B-Instruct/model.safetensors", "rb") as f:
    data = deserialize(f.read())
all_keys = {}
for key, _ in data:
    all_keys[key] = None
print(f"Time to call 'deserialize w/ reading keys': {time.time() - start}")
after = memory_usage()
print(f"Memory usage before: {before} MB")
print(f"Memory usage difference: {after - before} MB")
```

</p>
</details> 

It looks like anything that uses `safe_open` with the `get_tensor` abstraction actually does do lazy loading so perhaps that is a better way to go?

ankitageorge replied:
safe_open doesn't work with fsspec though. it's the original solution I had that isn't working.
I think not needing metadata would be a decent amount of refactor, especially since this is kind of an edge case with not many users, I'd rather keep the metadata, rather than making it optional and hacking the code everywhere else. 

The other option I found is to read metadata directly from the safetensors file like so, also a little hack-y but would save full load:
```
header_len_bytes = f.read(4)
 header_len = struct.unpack("i", header_len_bytes)[0]
f.read(4)
header_json = f.read(header_len)
metadata = json.loads(header_json)
print(metadata.keys())
   ```

joecummings replied:
It looks like fsspec *does* support an open operator: https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.open.

Currently, it's not implemented for the PyTorch version of fsspec, it seems like a quick addition (but n00b to fsspec here). 

Looking closer at the ``safe_open`` function, it appears that they actually have a `get_keys` function that would return exactly what we need here.

***

The above option is an interesting one for sure, but I don't feel comfortable hardcoding in header sizes. At the very least, is there a safetensor spec we could reference or even better some constants we could import to ensure we always read the exact right sizes?

ankitageorge replied:
If you look at the existing implementation before my changes, that might help with understanding, but the safe_open with the .keys() call is what I did, but it doesn't work to read from hf hub, only local fs. Safe_open has some extra custom deserialization specific for a safetensors file that's different than a regular open call, which is what gives us the keys() method, a regular open() call won't do that. The `create_stream` is essentially the same thing as `open`, so we have that for fsspec. But the issue is coming from the fact safe_open() only works for local filesystems.

joecummings replied:
Ahhhh, I see - the ``safe_open`` function in Rust expects the full file to be memory-mapped or passed to torch.from_file, which cannot be done via fsspec!! Thanks for bearing w/ me and helping me understand :)

Then I would rank my preference as the following:
1. Use the hack-y reading the header option, but only if there are concrete specs or imports we can use so that we always read the correct amount. If there's not then ...
2. Keep the current implementation, which loads up the entire file, but add a TODO or a comment saying that this is suboptimal and we should consider refactoring.

How does this align with your expectations?

ankitageorge replied:
I can just follow the logic that HF is already doing for option 1 https://github.com/2404589803/huggingface_hub/blob/a005e58f47005c16b82035da7a3c36745cade499/src/huggingface_hub/hf_api.py#L5308. The only issue is what if they change something from their end? I guess we can just cross that bridge if we get there?

joecummings replied:
https://github.com/huggingface/huggingface_hub/blob/df9f47bef32f090ae1adf2f10c8a6f6110b1d1a9/src/huggingface_hub/hf_api.py#L5819

joecummings replied:
Okay it looks like they also kind of hackily do it here. That's good enough for me. Can we just comment it well enough and maybe point to the spec?

joecummings replied:
Whoops I see you already commented that. Yep, looks good to me!!

joecummings replied:
This could be helpful to point to: https://huggingface.co/docs/safetensors/index#format

joecummings reviewed (approved):
Awesome stuff - thanks @ankitageorge !

ankitageorge committed:
Fix HF loading when there's no metadata file to work with fsspec (#152856)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/152856

HF loading when there is no metadata is an edge case for some users. We were previously calling safe_open(filename) to get the keys in the safetensors file, but this doesn't work with fsspec, when models have a different backend than local fs (ie. hf, s3 etc). This diff updates to open the file with fsspec.open() and then safetensors.deserialize() to get the keys

Test Plan: unit test and e2e test reading from hf

Differential Revision: D74181513

ankitageorge commented:
@pytorchmergebot merge -i
